hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/metrics/JobMetrics.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.core.metrics;
+
+import com.hazelcast.jet.impl.JobMetricsUtil;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Predicate;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * An immutable collection of job-specific metrics, pairs of metric names and sets of associated {@link Measurement}s.
+ *
+ * @since 3.2
+ */
+public final class JobMetrics implements IdentifiedDataSerializable {
+
+    private static final JobMetrics EMPTY = new JobMetrics(Collections.emptyMap());
+
+    private static final Collector<Measurement, ?, TreeMap<String, Set<Measurement>>> COLLECTOR = Collectors.groupingBy(
+            measurement -> measurement.getTag(MetricTags.METRIC),
+            TreeMap::new,
+            Collectors.mapping(identity(), toSet())
+    );
+
+    private Map<String, Set<Measurement>> metrics; //metric name -> set of measurements
```

Comment:
is there a reason to use Set here instead of List? We don't expect duplicate of same measurement anywhere?